### PR TITLE
Expand query for virtual product

### PR DESCRIPTION
**Pull request by emmaai**
- to reuse `_find_dss` in virtual product
- pin `datacube != 1.8.14` because of the issue in `1.8.14`

### CODE REVIEW

File: odc/stats/tasks.py [hunk 1]
```diff
@@ -151,18 +158,65 @@ def match_dss(groups, group_size):
...
+            dataset_filter = {}
+
+        # query by a list of products is not a "officially" supported feature
+        # but it is embedded in the code everywhere
+        # mark it for ref
+
+        product_list = re.split(r"\+|-", products)
+        product_list = list(filter(None, product_list))
+        group_size = len(re.findall(r"\+", products))
+
```
Comment (SpacemanPaul) [inline odc/stats/tasks.py:211]: I don't understand what's going on here.  What is the expected format of "products"?  That's some weird stuff going on with pluses and minuses.  

`products="prod1+prod2-prod3+prod4"` and `products="+++prod1+prod2-+-prod3++prod4"` both  yield the same 4 member `product_list`, but one givess a `group_size` of 4 and the other 7??  With no warning/error message?

It seems like the main purpose of `find_dss` is to expand a product string into a product list, so a little bit of documentation here would be helpful.

(It's still better than it was, mind you.)

File: odc/stats/tasks.py [hunk 2]
```diff
@@ -151,18 +158,65 @@ def match_dss(groups, group_size):
...
+            def map_fuse_func(x):
+                return fuse_ds(*x, product=fused_product)
+
+            dss = map(map_fuse_func, grouped_dss)
+            return dss
+        else:
+            return grouped_dss
+
+    @classmethod
+    def _find_dss(
```
Comment (SpacemanPaul) [inline odc/stats/tasks.py:173]: Having `_find_dss` and `_get_dss` side by side like this is potentially confusing.

File: setup.cfg [hunk 3]
```diff
@@ -22,7 +22,7 @@ install_requires =
     botocore
     click>=8.0.0
     dask
-    datacube
+    datacube<=1.8.13
```
Comment (SpacemanPaul) [inline setup.cfg]: As this issue has been fixed in develop and will therefore definitely be fixed in 1.8.15, maybe just require `!=1.8.14`

**Review by SpacemanPaul — CHANGES_REQUESTED** [review]
Good refactor, but a little more documentation (and/or validation) would be helpful.

Also this patch has only has like 15% coverage. A quick coverage run of  `find_dss` with mocks should be fairly quick to throw together.

File: odc/stats/tasks.py [hunk 4]
```diff
@@ -151,18 +158,85 @@ def match_dss(groups, group_size):
...
+        def sanitize_products_str(products_str):
+            """
+            split a string composed by product names into a list of product names
+            e.g., ga_ls8-ga_ls7 -> [ga_ls8, ga_ls7]
+            rules:
+            1. Any separator (`+/-`) at the start or the end is disregarded
+            2. Multiple same separators between the product names are treated as one
+            3. Multiple different separators between the product names is respected by left-right order
+            e.g., ga_ls8+-ga_ls7 -> separator is `+` as `+` proceeds `-` from left to right
+            """
```
Comment (SpacemanPaul) [inline odc/stats/tasks.py:204]: This is much better but I still don't understand the difference between + and - as a separator or what the significance of `group_size` is.

**Review by SpacemanPaul — APPROVED** [review]
The product string splitter could do with a just a bit more documentation (specifically around the "grouping" concept).  Otherwise looking great. Thanks Emma.

File: odc/stats/tasks.py (hunk 4 above)
Comment (emmaai) [reply odc/stats/tasks.py:204]: I'll throw in a small segment in readme in next pr. It's meant to support the concept of `juxtapose` and `collate` as well as the similar idea in `odc-stats` with the same set of code.

**Pull request merged**